### DISCUSSION TITLE
ONVIF and time sync?

### DISCUSSION THREAD

**Discussion by leaskovski** [blakeblackshear/frigate#8930] [Ideas]
So aparently you can perform time syncs using the ONVIF protocol.  With the newest beta build, we can setup ONVIF so we can get PTZ control of cameras.  With that integration, is there any chance we can also get time sync working as well?  Some of the cameras I am using seem to be crap and just plain ignore my DHCP NTP option, so the camera drifts out of sync over time.  Time sync over OVNIF would be great and solve this.

**Comment by hawkeye217** [comment]
I'm not sure if this is functionality that frigate or any NVR should provide (as NTP is the standard way that cameras will sync time).  However, it shouldn't be too hard to set up a cron job to that does what you're wanting to do.

I found [this post ](https://stackoverflow.com/questions/56638795/set-date-and-time-in-hikvision-camera-from-bash-or-rest)with some Python code that should probably do what you want.

**Comment by leaskovski** [comment]
I did get this working thanks to your post.  I did the following...

First we want to create a python virtual environment to store our python libs and code..
```
sudo  apt-get install python3-venv
python3 -m venv /usr/local/bin/onvif_timesync
source onvif_timesync/bin/activate
```
Install the required library that does the onvif functionality…
`pip install --upgrade onvif_zeep`
Now clone that library's git repo so we get the WSDL needed for the requests…
`git clone https://github.com/FalkTannhaeuser/python-onvif-zeep.git`
Now we can create our script…
`nano sync.py`
copy in, and update the parameters to ONVIFCamera(...)
```
	from onvif import ONVIFCamera
	import datetime
	
	mycam = ONVIFCamera(ip_address_string, port_number_int, onvif_user_string, onvif_password_string, '/usr/local/bin/onvif_timesync/python-onvif-zeep/wsdl')
	
	time_params = mycam.devicemgmt.create_type('SetSystemDateAndTime')
	time_params.DateTimeType = 'Manual'
	time_params.DaylightSavings = True
	
	TZ = {'TZ':datetime.datetime.now().astimezone().tzname()}
	time_params.TimeZone = TZ
	
	today = datetime.datetime.now()
	
	Date = {'Year':today.year,'Month':today.month,'Day':today.day}
	Time = {'Hour':today.hour,'Minute':today.minute,'Second':today.second}
	
	UTC = {'Date':Date,'Time':Time}
	time_params.UTCDateTime = UTC
	
	mycam.devicemgmt.SetSystemDateAndTime(time_params)
```
Exit the virtual environment…
`deactivate`
Run the script when we want…
`/usr/local/bin/onvif_timesync/bin/python /usr/local/bin/onvif_timesync/sync.py`

**Reply by leaskovski** [reply]
I did convert this into a HACS plugin... https://github.com/leaskovski/onvif_timesync

**Comment by alesgenova** [comment]
Based on this, I created a small service I add to my `frigate` docker compose to periodically sync the time of my ONVIF compatible cameras to a custom NTP server. This way the cameras now always have accurate time, and are still completely cut off from the internet.

If someone has the similar use case, you can use it like this:
```yaml
services:
  onvif-time:
    image: ghcr.io/alesgenova/onvif-time:dev
    volumes:
      - /path/to/config.json:/config/config.json
    environment:
      TZ: "America/New_York"
      ONVIF_TIME_SCHEDULE: "0 2 * * *"
      ONVIF_TIME_STARTUP_SYNC: True
    restart: unless-stopped
```

And the configuration file looks like this:
```json
{
  "ntp": {
    "host": "pool.ntp.org",
    "port": 123,
    "timeout": 5
  },
  "cameras": {
    "front_camera": {
      "host": "10.0.0.2",
      "port": 12345,
      "user": "user",
      "password": "password"
    },
    "back_camera": {
      "host": "10.0.0.3",
      "port": 12345,
      "user": "user",
      "password": "password"
    }
  }
}
```

Full source:
https://github.com/alesgenova/onvif-time
